### PR TITLE
chore: implement "env upgrade" Execute

### PR DESCRIPTION
**Pull request by efekarakus**
With this change, we're done with the implementation of the "env upgrade" command.
Here is an overview of the decision tree on how environments are upgraded:
```
Is environment template
├── Not a legacy template and there is a newer version?
│   └── Upgrade easily
├── A legacy template but with default VPC configuration?
│   └── Upgrade legacy template 
└── A legacy template but with modified VPC?
    └── Is the VPC configuration in SSM?
        └── Upgrade legacy template 
    └── Not in SSM so throw an error
```

_By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice._

### CODE REVIEW

File: internal/pkg/cli/env_upgrade.go [hunk 1]
```diff
@@ -183,6 +241,103 @@ Are you using the latest version of AWS Copilot?`, env, deploy.LatestEnvTemplate
...
+		}
+		lbWebServiceNames = append(lbWebServiceNames, svc.Name)
+	}
+	return lbWebServiceNames, nil
+}
+
+func (o *envUpgradeOpts) upgradeLegacyEnvironmentWithVPCOverrides(upgrader legacyEnvUpgrader, conf *config.Environment,
+	fromVersion, toVersion string, albWorkloads []string) error {
+	if conf.CustomConfig != nil {
+		err := upgrader.UpgradeLegacyEnvironment(&deploy.CreateEnvironmentInput{
```
Comment (iamhopaul123) [inline internal/pkg/cli/env_upgrade.go]:
```suggestion
		if err := upgrader.UpgradeLegacyEnvironment(&deploy.CreateEnvironmentInput{...}}; err != nil {
```

File: internal/pkg/cli/env_upgrade.go [hunk 2]
```diff
@@ -55,6 +69,10 @@ func newEnvUpgradeOpts(vars envUpgradeVars) (*envUpgradeOpts, error) {
 
 		store: store,
 		sel:   selector.NewSelect(prompt.New(), store),
+		envTemplate: stack.NewEnvStackConfig(&deploy.CreateEnvironmentInput{
```
Comment (iamhopaul123) [inline internal/pkg/cli/env_upgrade.go]: Should this be named as
```suggestion
		legacyEnvTemplater: stack.NewEnvStackConfig(&deploy.CreateEnvironmentInput{
```
What do you think?

File: internal/pkg/cli/env_upgrade.go [hunk 3]
```diff
@@ -183,6 +241,103 @@ Are you using the latest version of AWS Copilot?`, env, deploy.LatestEnvTemplate
...
+		}
+		return nil
+	}
+	return o.upgradeLegacyEnvironmentWithVPCOverrides(upgrader, conf, fromVersion, toVersion, albWorkloads)
+}
+
+func (o *envUpgradeOpts) isDefaultLegacyTemplate(cfn envTemplater, appName, envName string) (bool, error) {
+	defaultLegacyEnvTemplate, err := o.envTemplate.Template()
+	if err != nil {
+		return false, fmt.Errorf("generate environment template %s: %v", deploy.LegacyEnvTemplateVersion, err)
```
Comment (iamhopaul123) [inline internal/pkg/cli/env_upgrade.go]:
```suggestion
		return false, fmt.Errorf("generate default legacy environment template: %v", err)
```

File: internal/pkg/cli/env_upgrade.go [hunk 4]
```diff
@@ -37,12 +47,16 @@ type envUpgradeVars struct {
...
+	sel         appEnvSelector
+	envTemplate templater
+	prog        progress
 
 	// Constructors for clients that can be initialized only at runtime.
 	// These functions are overriden in tests to provide mocks.
-	newEnvVersionGetter func(app, env string) (versionGetter, error)
+	newEnvVersionGetter    func(app, env string) (versionGetter, error)
+	newTemplateUpgrader    func(conf *config.Environment) (envTemplateUpgrader, error)
+	newEnvOutputsDescriber func(app, env string) (envVPCDescriber, error)
```
Comment (iamhopaul123) [inline internal/pkg/cli/env_upgrade.go]: Do we use this anywhere 🤔

File: internal/pkg/describe/env.go [hunk 5]
```diff
@@ -120,6 +121,41 @@ func (d *EnvDescriber) Version() (string, error) {
...
 
+// EnvironmentVPC holds the ID of the environment's VPC configuration.
+type EnvironmentVPC struct {
+	ID               string
+	PublicSubnetIDs  []string
+	PrivateSubnetIDs []string
+}
+
+// EnvironmentVPC returns the ID of the VPC and its subnets by reading the outputs of the environment cloudformation stack.
+func (d *EnvDescriber) EnvironmentVPC() (*EnvironmentVPC, error) {
```
Comment (iamhopaul123) [inline internal/pkg/describe/env.go]: Do we use this function?

**Review by iamhopaul123 — COMMENTED** [review]
This is amazing! Looking forward to the env upgrade feature!

File: internal/pkg/cli/env_upgrade.go (hunk 2 above)
Comment (efekarakus) [reply internal/pkg/cli/env_upgrade.go]: 👍 great suggestion!

File: internal/pkg/cli/env_upgrade.go (hunk 4 above)
Comment (efekarakus) [reply internal/pkg/cli/env_upgrade.go]: 🤦 lol i used it while experimenting if there was a way of reading the VPC configuration by checking the outputs of the cloudformation stack, and then forgot to remove the code.

Thanks for pointing it out!

File: internal/pkg/describe/env.go (hunk 5 above)
Comment (efekarakus) [reply internal/pkg/describe/env.go]: 🤦 same as above, it was code that I had written while experimenting with the cmd. Deleted it :P

**Review by iamhopaul123 — APPROVED** [review]
:shipit:

**Review by huanjani — APPROVED** [review]

**Pull request merged**